File: py/base_api_utils.py
# -*- coding: utf-8 -*-

from math import exp
import requests


class BaseApiUtils:

    def __init__(self, cfg):
        self.cfg = cfg

    # GLOBAL VALID TYPES
    VALID_MATCH_TYPES = ['exact', 'contains']
    VALID_BODY_TYPES = ['json', 'text']

    # shared requests session object for all sessions
    req_sess = requests.Session()

    def __send_request_without_payload(self, method, url, params):
        if method.lower() == 'get':
            rsp = self.req_sess.get(url=url, params=params)
        elif method.lower() == 'post':
            rsp = self.req_sess.post(url=url, params=params)
        elif method.lower() == 'put':
            rsp = self.req_sess.put(url=url, params=params)
        elif method.lower() == 'patch':
            rsp = self.req_sess.patch(url=url, params=params)
        elif method.lower() == 'delete':
            rsp = self.req_sess.delete(url=url, params=params)
        elif method.lower() == 'head':
            rsp = self.req_sess.head(url=url, params=params)
        elif method.lower() == 'options':
            rsp = self.req_sess.options(url=url, params=params)
        else: # unknown method
            raise AttributeError("unknown method type: '{0}'".format(str(method)))

        return rsp

    def __send_request_with_payload_as_json(self, method, url, payload, params):
        if method.lower() == 'get':
            rsp = self.req_sess.get(url=url, json=payload, params=params)
        elif method.lower() == 'post':
            rsp = self.req_sess.post(url=url, json=payload, params=params)
        elif method.lower() == 'put':
            rsp = self.req_sess.put(url=url, json=payload, params=params)
        elif method.lower() == 'patch':
            rsp = self.req_sess.patch(url=url, json=payload, params=params)
        elif method.lower() == 'delete':
            rsp = self.req_sess.delete(url=url, json=payload, params=params)
        elif method.lower() == 'head':
            rsp = self.req_sess.head(url=url, json=payload, params=params)
        elif method.lower() == 'options':
            rsp = self.req_sess.options(url=url, json=payload, params=params)
        else: # unknown method
            raise AttributeError("unknown method type: '{0}'".format(str(method)))

        return rsp

    def __send_request_with_payload_as_data(self, method, url, payload, params):
        if method.lower() == 'get':
            rsp = self.req_sess.get(url=url, data=payload, params=params)
        elif method.lower() == 'post':
            rsp = self.req_sess.post(url=url, data=payload, params=params)
        elif method.lower() == 'put':
            rsp = self.req_sess.put(url=url, data=payload, params=params)
        elif method.lower() == 'patch':
            rsp = self.req_sess.patch(url=url, data=payload, params=params)
        elif method.lower() == 'delete':
            rsp = self.req_sess.delete(url=url, data=payload, params=params)
        elif method.lower() == 'head':
            rsp = self.req_sess.head(url=url, data=payload, params=params)
        elif method.lower() == 'options':
            rsp = self.req_sess.options(url=url, data=payload, params=params)
        else: # unknown method
            raise AttributeError("unknown method type: '{0}'".format(str(method)))

        return rsp

    def __call_api(self, method, url, payload, params=None, payload_as_json=True):
        if payload is not None:
            if payload_as_json: # send request payload as json
                rsp = self.__send_request_with_payload_as_json(
                    method=method,
                    url=url,
                    payload=payload,
                    params=params
                )
            else: # send request payload as data
                rsp = self.__send_request_with_payload_as_data(
                    method=method,
                    url=url,
                    payload=payload,
                    params=params
                )
        else: # send request without a payload
            rsp = self.__send_request_without_payload(
                method=method,
                url=url,
                params=params
            )

        return rsp

    def check_response_status_code(self, rsp, expected_status_code):
        bool = False
        actual_status_code = rsp.status_code
        bool = (str(actual_status_code) == str(expected_status_code))

        if bool:
            msg = "success"
        else: 
            msg = (
                "for {0} request against url: {1}\n"
                "expected response status code not found\n"
                "  actual status code   : {2}\n"
                "  expected status code : {3}\n"
                "response body/text: {4}".format(
                    str(rsp.request.method),
                    str(rsp.request.url),
                    str(actual_status_code),
                    str(expected_status_code),
                    str(rsp.text)
                )
            )

        return bool, msg

    def check_response_body(self, rsp, expected_body, body_type='json', match_type='exact'):
        bool = False
        try:
            actual_body = rsp.json()
        except Exception as e:
            actual_body = rsp.text
        if match_type.lower() == 'contains':
            bool = (str(expected_body) in str(actual_body))
        elif match_type.lower() == 'exact':
            if body_type.lower() == 'json':
                bool = (expected_body == actual_body)
            elif body_type.lower() == 'text':
                bool = (str(expected_body) == str(actual_body))
            else: # unknown body_type
                raise AttributeError(
                    "unknown body type: '{0}'\n"
                    "valid body types are: '{1}'".format(
                        str(body_type),
                        str(self.VALID_BODY_TYPES)
                    )
                )
        else: # unknown match_type
            raise AttributeError(
                "unknown match type: '{0}'\n"
                "valid match types are: '{1}'".format(
                    str(match_type),
                    str(self.VALID_MATCH_TYPES)
                )
            )
        
        if bool:
            msg = "success"
        else:
            msg = (
                "for {0} request against url: {1}\n"
                "via '{2}' match for {3} type, expected response body not found\n"
                "  actual body   : {4}\n"
                "  expected body : {5}\n".format(
                    str(rsp.request.method),
                    str(rsp.request.url),
                    str(match_type),
                    str(body_type),
                    str(actual_body),
                    str(expected_body)
                )
            )

        return bool, msg

    def check_response_headers(self, rsp, expected_header_key, expected_header_value):
        bool = False
        actual_headers = rsp.headers
        if actual_headers.get(expected_header_key) != None:
            actual_header_value = actual_headers[expected_header_key]
            bool = (str(expected_header_value) in str(actual_header_value))
        
        if bool:
            msg = "success"
        else:
            msg = (
                "for {0} request against url: {1}\n"
                "via 'contains' match, expected response headers not found\n"
                "  actual headers              : {2}\n"
                "  expected header (key:value) : {3}:{4}\n"
                "response body/text: {5}".format(
                    str(rsp.request.method),
                    str(rsp.request.url),
                    str(actual_headers),
                    str(expected_header_key),
                    str(expected_header_value),
                    str(rsp.text)
                )
            )

        return bool, msg
    
    def check_response_content_type(self, rsp, expected_header_value):
        return self.check_response_headers(rsp=rsp, expected_header_key='content-type', expected_header_value=expected_header_value)

    def check_response_allow_methods(self, rsp, expected_header_value):
        return self.check_response_headers(rsp=rsp, expected_header_key='allow', expected_header_value=expected_header_value)

    def delete_it(self, url, params=None):
        rsp = self.__call_api(method='delete', url=url, payload=None, params=params)
        return rsp

    def get_it(self, url, params=None):
        rsp = self.__call_api(method='get', url=url, payload=None, params=params)
        return rsp

    def head_it(self, url, params=None):
        rsp = self.__call_api(method='head', url=url, payload=None, params=params)
        return rsp

    def options_it(self, url, params=None):
        rsp = self.__call_api(method='options', url=url, payload=None, params=params)
        return rsp

    def patch_it(self, url, params=None):
        rsp = self.__call_api(method='patch', url=url, payload=None, params=params)
        return rsp

    def post_it(self, url, params=None):
        rsp = self.__call_api(method='post', url=url, payload=None, params=params)
        return rsp

    def put_it(self, url, params=None):
        rsp = self.__call_api(method='put', url=url, payload=None, params=params)
        return rsp

File: rb/base_api_utils.rb
# encoding: utf-8

module BaseApiUtils

    @@valid_match_types = ['exact', 'contains']
    @@valid_body_types = ['json', 'text']

    def send_request_without_payload(method, url, params)
        begin
            case method.to_s.downcase
            when 'get', 'post', 'put', 'patch', 'delete', 'head', 'options'
                rsp = RestClient::Request.execute(
                    method: :"#{method.to_s.downcase}",
                    url: url,
                    params: params) { |response|
                        case response.code
                        when 301, 302, 307
                            response.follow_redirection
                        else
                            response.return!
                        end # case response block
                } # execute block
            else # unknown value
                raise ArgumentError, "unknown method type: #{method.to_s}"
            end # case method block
        rescue RestClient::ExceptionWithResponse => err
            rsp = err.response
        end # begin
        return rsp
    end
    private :send_request_without_payload

    def send_request_with_payload(method, url, payload, params)
        begin
            case method.to_s.downcase
            when 'get', 'post', 'put', 'patch', 'delete', 'head', 'options'
                rsp = RestClient::Request.execute(
                    method: :"#{method.to_s.downcase}",
                    url: url,
                    payload: payload,
                    params: params) { |response|
                        case response.code
                        when 301, 302, 307
                            response.follow_redirection
                        else
                            response.return!
                        end # case response block
                } # execute block
            else # unknown value
                raise ArgumentError, "unknown method type: #{method.to_s}"
            end # case method block
        rescue RestClient::ExceptionWithResponse => err
            rsp = err.response
        end # begin
        return rsp
    end
    private :send_request_with_payload

    def send_request_with_payload_as_json(method, url, payload, params)
        begin
            case method.to_s.downcase
            when 'delete', 'get', 'head', 'options', 'patch', 'post', 'put'
                rsp = RestClient::Request.execute(
                    method: :"#{method.to_s.downcase}",
                    url: url,
                    payload: payload.to_json,
                    params: params,
                    headers: {content_type: :json}) { |response|
                        case response.code
                        when 301, 302, 307
                            response.follow_redirection
                        else
                            response.return!
                        end # case response block
                } # execute block
            else # unknown value
                raise ArgumentError, "unknown method type: #{method.to_s}"
            end # case method block
        rescue RestClient::ExceptionWithResponse => err
            rsp = err.response
        end # begin
        return rsp
    end
    private :send_request_with_payload_as_json

    def call_api(method, url, payload=nil, params=nil, payload_as_json=true)
        unless payload.nil?
            if payload_as_json # send request payload as a json
                rsp = send_request_with_payload_as_json(
                    method=method,
                    url=url,
                    payload=payload,
                    params=params
                )
            else # send request with payload
                rsp = send_request_with_payload(
                    method=method,
                    url=url,
                    payload=payload,
                    params=params
                )
            end
        else # send request without a payload
            rsp = send_request_without_payload(
                method=method,
                url=url,
                params=params
            )
        end
        return rsp
    end
    private :call_api

    def check_response_status_code(rsp, expected_status_code)
        bool = false
        actual_status_code = rsp.code
        bool = (actual_status_code.to_s == expected_status_code.to_s)

        if bool
            msg = "success"
        else
            msg = "for #{rsp.request.method} request against url: #{rsp.request.url}\n" \
            "expected response status code not found\n" \
            "  actual status code   : #{actual_status_code}\n" \
            "  expected status code : #{expected_status_code}\n" \
            "response body: #{rsp.body}"
        end

        return bool, msg
    end

    def check_response_body(rsp, expected_body, match_type='exact')
        bool = false
        begin
            actual_body = JSON.parse(rsp.body).to_json
        rescue JSON::ParserError
            actual_body = rsp.body
        end
        if match_type.to_s.downcase == 'contains'
            bool = (expected_body.to_s.include? actual_body.to_s)
        elsif match_type.to_s.downcase == 'exact'
            bool = (expected_body.to_s == actual_body.to_s)
        else # unknown match_type
            raise ArgumentError, "unknown match type: #{match_type}\n" \
                                 "valid match types are: #{@@valid_match_types}"
        end

        if bool
            msg = "success"
        else
            msg = "for #{rsp.request.method} request against url: #{rsp.request.url}\n" \
            "via #{match_type} match, expected response body not found\n" \
            "  actual body   : #{actual_body}\n" \
            "  expected body : #{expected_body}\n"
        end

        return bool, msg 
    end

    def check_response_headers(rsp, expected_header_key, expected_header_value)
        bool = false
        actual_headers = rsp.headers
        unless actual_headers[expected_header_key.to_sym].nil?
            actual_header_value = actual_headers[expected_header_key.to_sym]
            bool = (actual_header_value.to_s.include? expected_header_value.to_s)
        end

        if bool
            msg = "success"
        else
            msg = "for #{rsp.request.method} request against url: #{rsp.request.url}\n" \
            "via 'contains' match, expected response headers not found\n" \
            "  actual headers               : #{actual_headers}\n" \
            "  expected headers (key:value) : #{expected_header_key}:#{expected_header_value}\n" \
            "response body: #{rsp.body}"
        end

        return bool, msg 
    end

    def check_response_content_type(rsp, expected_header_value)
        return check_response_headers(rsp=rsp, expected_header_key='content_type', expected_header_value=expected_header_value)
    end

    def check_response_allow_methods(rsp, expected_header_value)
        return check_response_headers(rsp=rsp, expected_header_key='allow', expected_header_value=expected_header_value)
    end

    def delete_it(url, params=nil)
        rsp = call_api(method='delete', url=url, payload=nil, params=params)
        return rsp
    end

    def get_it(url, params=nil)
        rsp = call_api(method='get', url=url, payload=nil, params=params)
        return rsp
    end

    def head_it(url, params=nil)
        rsp = call_api(method='head', url=url, payload=nil, params=params)
        return rsp
    end

    def options_it(url, params=nil)
        rsp = call_api(method='options', url=url, payload=nil, params=params)
        return rsp
    end

    def patch_it(url, params=nil)
        rsp = call_api(method='patch', url=url, payload=nil, params=params)
        return rsp
    end

    def post_it(url, params=nil)
        rsp = call_api(method='post', url=url, payload=nil, params=params)
        return rsp
    end

    def put_it(url, params=nil)
        rsp = call_api(method='put', url=url, payload=nil, params=params)
        return rsp
    end

end # module end
